cpp/examples/kmeans/kmeans_example.cpp

```diff
@@ -68,41 +68,6 @@ bool get_arg(char **begin, char **end, const std::string &arg) {
   return false;
 }
 
-class cachingDeviceAllocator : public ML::deviceAllocator {
```

Comment:
I think most changes are just moving around headers... out of curiosity, why do these cachingDeviceAllocator blocks get removed?

Reply:
If I'm not mistaken its because they were part of the examples instead of inside cuML C++ library itself

Reply:
that's true. `cachingDeviceAllocator` is a wrapper over the corresponding allocator inside cub. We have cub as our dependency anyways. So, made little sense to keep it inside examples. Thus, moved this inside our include folders.

It also helped resolve a cuml -> ml-prims -> cuml circular dependency on header files.